### ISSUE TITLE
How to use it with the plain psr-15?

### ISSUE THREAD

**Issue by erickskrauch** [dunglas/frankenphp#1179]
I can't make it running. I'm getting the error
```
PHP Fatal error:  Maximum call stack size of 8339456 bytes (zend.max_allowed_stack_size - zend.reserved_stack_size) reached during compilation. Try splitting expression in /home/erickskrauch/projects/elyby/skins-renderer/src/index.php on line 2  {"syslog_level": "notice"}
```

While the line 2 is `declare(strict_types=1);` :upside_down_face: 

```php
<?php
declare(strict_types=1);

use Ely\SkinsRenderer\Application;
use Psr\Http\Server\RequestHandlerInterface;

require __DIR__ . '/../vendor/autoload_runtime.php';

return static function(array $context): RequestHandlerInterface {
    return new Application();
};
```

I'm trying to run both on Linux native and in Docker.

**Comment by withinboredom** [comment]
There's not much we can help you with without the `phpinfo` dump you were prompted to enter when filing the PR. What version of PHP are you running, what extensions, etc. This could be a known bug in php, or a bug in frankenphp, but without considerably more information, it is hard for us to help.

**Comment by withinboredom** [comment]
Does it still crash without strict types? FWIW, there's no reason for strict types here (you aren't calling any functions with scalars), but that's beside the point. I'm just curious if that is the issue.

**Comment by erickskrauch** [comment]
No, commenting out the line with strict types makes it crash on the next meaningful row (`require`). Commenting out `require` makes it crash on the `return`. But there is no error about types or something. It is always the `Maximum call stack size...`.

**Comment by dunglas** [comment]
Can you try to increase the stack size (see the tip at the end of this page): https://frankenphp.dev/docs/compile/#using-xcaddy

If this doesn't fix the issue, you likely have an infinite loop somewhere in your code.

**Comment by erickskrauch** [comment]
Yeah, it's the most obvious thought. But I don't see any loops in my code or autoloading rules. I've pushed my code. If you have time and will, please take a look at it.
https://github.com/elyby/skins-renderer/tree/frankenphp

**Comment by withinboredom** [comment]
Am I reading the dockerfile correctly that you are using a non-zts build of php? I can't find `ppm.json` so I have no idea what is going on or what you are trying to accomplish. What is in `vendor/autoload_runtime.php`? If you are using frankenphp, I see you starting up a worker, but no [worker script](https://frankenphp.dev/docs/worker/#custom-apps).

**Comment by erickskrauch** [comment]
> Am I reading the dockerfile correctly that you are using a non-zts build of php? I can't find ppm.json so I have no idea what is going on or what you are trying to accomplish.

Sorry for the confusion. This is a WIP project state, so this Dockerfile isn't relevant to what I'm trying to accomplish. I'm trying to run the FrankenPHP on my computer in the most direct way possible:
```sh
env APP_RUNTIME=Runtime\\PsrGuzzle\\Runtime frankenphp php-server --worker ./src/index.php -v -a
```

And this is how I get those errors I reported initially.

> What is in vendor/autoload_runtime.php?

I believe this file is generated by the `symfony/runner`. Usage of this file was suggested in the [`runtime/frankenphp-symfony`](https://github.com/php-runtime/frankenphp-symfony)'s README. But since my project doesn't use the Symfony (it is overkill for what I need to implement), I'm looking for the solution with a pure PSR-15.

**Comment by withinboredom** [comment]
I don't believe anyone has built a pure PSR-15 worker script for FrankenPHP (I haven't searched for it, though). To use worker mode, you need a worker script: something to boot up your project before any requests start (if needed), and then call `frankenphp_handle_request` to handle the request. In that callback would be where you configure the request/response before calling your framework code.

see: [custom apps](https://frankenphp.dev/docs/worker/#custom-apps)

**Comment by erickskrauch** [comment]
Then consider it a feature request. Adapting PSR-15 will make integration very easy for everyone. The current symfony runner support for Symfony-only is too specific. Making universal support for all runner runtimes would greatly expand the applicability of FrankenPHP.

**Comment by withinboredom** [comment]
FrankenPHP is designed to be flexible and support any runtime you can imagine and isn't tied to Symfony or Laravel. As these are the two most popular PHP frameworks in the world, we have a lot of documentation helping people out with them.

If you haven't already, I highly recommend checking out the documentation I shared earlier. It outlines how FrankenPHP is not limited to Symfony or Laravel but can work seamlessly with a variety of setups, including PSR-15. Let us know if you have specific questions after reviewing it—we’re happy to help!

**Comment by gggeek** [comment]
@erickskrauch I have used Nyholm\Psr7\Factory\Psr17Factory and Nyholm\Psr7Server\ServerRequestCreator (a patched version of that one, as it has a few bugs) to achieve what I think you are requesting:

In index.php, before the worker-mode loop:
```php
    $psr17Factory = new Psr17Factory();
    $creator = new ServerRequestCreator(
        $psr17Factory, // ServerRequestFactory
        $psr17Factory, // UriFactory
        $psr17Factory, // UploadedFileFactory
        $psr17Factory  // StreamFactory
    );
```
and in the handler function passed in to `frankenphp_handle_request`:
```php
$serverRequest = $creator->fromGlobals();
```

It does not look like a huge burden. It could possibly be added to the current docs page about worker mode, as a possible implementation for psr-15-based apps?

Or are you more interested in the possible speedup which could be achieved by having go code inject the ServerRequestInterface object into the php script, akin to what Swoole does in https://github.com/openswoole/openswoole/blob/master/core/src/Helper.php, mapping the `\OpenSwoole\HTTP\Request` object into a psr ServerRequest one?

ps: full code sample (with lots of unrelated stuff) at https://github.com/gggeek/yet-another-docker-socket-proxy/blob/main/public/index.php

**Comment by dunglas** [comment]
Doc PR opened: https://github.com/php/frankenphp/pull/2539 (PSR-15 worker example).
